target-chains/near/receiver/src/governance.rs

```diff
@@ -34,162 +29,221 @@ use {
+#[strum_discriminants(derive(num_derive::ToPrimitive, num_derive::FromPrimitive))]
+#[strum_discriminants(name(GovernanceActionId))]
 #[serde(crate = "near_sdk::serde")]
 pub enum GovernanceAction {
-    ContractUpgrade([u8; 32]),
-    SetDataSources(Vec<Source>),
-    SetGovernanceSource(Source),
-    SetStalePriceThreshold(u64),
-    SetUpdateFee(u64),
+    UpgradeContract { codehash: [u8; 32] },
+    AuthorizeGovernanceDataSourceTransfer { claim_vaa: Vec<u8> },
+    SetDataSources { data_sources: Vec<Source> },
+    SetFee { val: u64, expo: u64 },
+    SetValidPeriod { valid_seconds: u64 },
+    RequestGovernanceDataSourceTransfer { governance_data_source_index: u32 },
 }
 
 impl GovernanceAction {
-    pub fn id(&self) -> ActionId {
-        match self {
-            GovernanceAction::ContractUpgrade(_) => ActionId::ContractUpgrade,
-            GovernanceAction::SetDataSources(_) => ActionId::SetDataSources,
-            GovernanceAction::SetGovernanceSource(_) => ActionId::SetGovernanceSource,
-            GovernanceAction::SetStalePriceThreshold(_) => ActionId::SetStalePriceThreshold,
-            GovernanceAction::SetUpdateFee(_) => ActionId::SetUpdateFee,
+    /// Implements a `deserialize` method for the `GovernanceAction` enum using `nom` to
+    /// deserialize the payload. The use of `nom` gives us parser safety, error handling, full
+    /// buffer consumption, and a more readable implementation while staying efficient.
+    pub fn deserialize(input: impl AsRef<[u8]>) -> Result<Self, Error> {
+        use nom::{
+            bytes::complete::take,
+            combinator::all_consuming,
+            multi::many0,
+            number::complete::{
+                be_u16,
+                be_u32,
+                be_u64,
+                be_u8,
+            },
+        };
+
+        let input = input.as_ref();
+
+        // Verify Governance header is as expected so we can bail to avoid more parsing.
+        let (input, magic) = take(4usize)(input)?;
+        if magic != GOVERNANCE_MAGIC {
+            return Err(Error::InvalidGovernanceModule);
         }
-    }
 
-    pub fn deserialize(data: &[u8]) -> Result<Self, Error> {
-        let mut cursor = std::io::Cursor::new(data);
-        let magic = cursor.read_u32::<BigEndian>()?;
-        let module = cursor.read_u8()?;
-        let action = cursor.read_u8()?.try_into()?;
-        let target = cursor.read_u16::<BigEndian>()?;
-
-        assert!(module == GOVERNANCE_MODULE);
-        assert!(target == 0 || target == u16::from(WormholeChain::Near));
-        assert!(magic == u32::from_le_bytes(GOVERNANCE_MAGIC));
-
-        Ok(match action {
-            ActionId::ContractUpgrade => {
-                let mut hash = [0u8; 32];
-                cursor.read_exact(&mut hash)?;
-                Self::ContractUpgrade(hash)
-            }
+        // Verify Target is this contract.
+        let (input, target) = be_u8(input)?;
+        if target != GOVERNANCE_MODULE {
```

Comment:
`GovernanceModule::Target as u8` ? (I think the constant GOVERNANCE_MODULE is redundant now)